### PR TITLE
Support FunctionArguments in HJSON configuration

### PR DESCRIPTION
**Pull request by rishi-aga**
One of the PR that Resolves https://github.com/yahoo/elide/discussions/1829

## Description
Support arguments at dimension, measure, and table level.
Support SQL helper syntax

## How Has This Been Tested?
Additional tests included

## License
I confirm that this contribution is made under an Apache 2.0 license and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

File: elide-model-config/src/main/java/com/yahoo/elide/modelconfig/DraftV4LibraryWithElideFormatAttr.java [hunk 1]
```diff
@@ -36,6 +38,7 @@ public DraftV4LibraryWithElideFormatAttr() {
         LibraryBuilder builder = DraftV4Library.get().thaw();
 
         builder.addFormatAttribute(ElideFieldNameFormatAttr.FORMAT_NAME, new ElideFieldNameFormatAttr());
+        builder.addFormatAttribute(ElideArgumentNameFormatAttr.FORMAT_NAME, new ElideArgumentNameFormatAttr());
         builder.addFormatAttribute(ElideCardinalityFormatAttr.FORMAT_NAME, new ElideCardinalityFormatAttr());
         builder.addFormatAttribute(ElideFieldTypeFormatAttr.FORMAT_NAME, new ElideFieldTypeFormatAttr());
         builder.addFormatAttribute(ElideGrainTypeFormatAttr.FORMAT_NAME, new ElideGrainTypeFormatAttr());
```
Comment (wcekan) [inline elide-model-config/src/main/java/com/yahoo/elide/modelconfig/DraftV4LibraryWithElideFormatAttr.java:44]: For builders it reads better to use running adds...
```suggestion
        library = DraftV4Library.get().thaw()

                .addFormatAttribute(ElideFieldNameFormatAttr.FORMAT_NAME, new ElideFieldNameFormatAttr())
                .addFormatAttribute(ElideArgumentNameFormatAttr.FORMAT_NAME, new ElideArgumentNameFormatAttr())
                .addFormatAttribute(ElideCardinalityFormatAttr.FORMAT_NAME, new ElideCardinalityFormatAttr())
                .addFormatAttribute(ElideFieldTypeFormatAttr.FORMAT_NAME, new ElideFieldTypeFormatAttr())
                .addFormatAttribute(ElideGrainTypeFormatAttr.FORMAT_NAME, new ElideGrainTypeFormatAttr())
```
etc

File: elide-model-config/src/main/java/com/yahoo/elide/modelconfig/DraftV4LibraryWithElideFormatAttr.java [hunk 2]
```diff
@@ -51,6 +54,7 @@ public DraftV4LibraryWithElideFormatAttr() {
 
         builder.addKeyword(new ValidateDimPropertiesKeyword().getKeyword());
         builder.addKeyword(new ValidateTimeDimPropertiesKeyword().getKeyword());
+        builder.addKeyword(new ValidateArgsPropertiesKeyword().getKeyword());
 
         library = builder.freeze();
```
Comment (wcekan) [inline elide-model-config/src/main/java/com/yahoo/elide/modelconfig/DraftV4LibraryWithElideFormatAttr.java:59]:
```suggestion

                .addKeyword(new ValidateDimPropertiesKeyword().getKeyword());
                .addKeyword(new ValidateTimeDimPropertiesKeyword().getKeyword());
                .addKeyword(new ValidateArgsPropertiesKeyword().getKeyword());

                 .freeze();
```

File: elide-model-config/src/main/java/com/yahoo/elide/modelconfig/MessageBundleWithElideMessages.java [hunk 3]
```diff
@@ -40,13 +42,16 @@ public MessageBundleWithElideMessages() {
...
         msgSourceBuilder.put(ElideFieldTypeFormatAttr.TYPE_KEY, ElideFieldTypeFormatAttr.TYPE_MSG);
         msgSourceBuilder.put(ElideGrainTypeFormatAttr.TYPE_KEY, ElideGrainTypeFormatAttr.TYPE_MSG);
         msgSourceBuilder.put(ElideJoinTypeFormatAttr.TYPE_KEY, ElideJoinTypeFormatAttr.TYPE_MSG);
         msgSourceBuilder.put(ElideJoinKindFormatAttr.TYPE_KEY, ElideJoinKindFormatAttr.TYPE_MSG);
         msgSourceBuilder.put(ElideTimeFieldTypeFormatAttr.TYPE_KEY, ElideTimeFieldTypeFormatAttr.TYPE_MSG);
         msgSourceBuilder.put(ElideNameFormatAttr.FORMAT_KEY, ElideNameFormatAttr.FORMAT_MSG);
+        msgSourceBuilder.put(ElideNameFormatAttr.NAME_KEY, ElideNameFormatAttr.NAME_MSG);
         msgSourceBuilder.put(ElideRSQLFilterFormatAttr.FORMAT_KEY, ElideRSQLFilterFormatAttr.FORMAT_MSG);
         msgSourceBuilder.put(JavaClassNameWithExtFormatAttr.FORMAT_KEY, JavaClassNameWithExtFormatAttr.FORMAT_MSG);
         msgSourceBuilder.put(JavaClassNameFormatAttr.FORMAT_KEY, JavaClassNameFormatAttr.FORMAT_MSG);
```
Comment (wcekan) [inline elide-model-config/src/main/java/com/yahoo/elide/modelconfig/MessageBundleWithElideMessages.java:57]: same here
```suggestion
                .put(ElideFieldNameFormatAttr.FORMAT_KEY, ElideFieldNameFormatAttr.FORMAT_MSG)
                .put(ElideFieldNameFormatAttr.NAME_KEY, ElideFieldNameFormatAttr.NAME_MSG)
                .put(ElideArgumentNameFormatAttr.FORMAT_KEY, ElideArgumentNameFormatAttr.FORMAT_MSG)
                .put(ElideArgumentNameFormatAttr.NAME_KEY, ElideArgumentNameFormatAttr.NAME_MSG)
                .put(ElideCardinalityFormatAttr.TYPE_KEY, ElideCardinalityFormatAttr.TYPE_MSG)
                .put(ElideFieldTypeFormatAttr.TYPE_KEY, ElideFieldTypeFormatAttr.TYPE_MSG)
                .put(ElideGrainTypeFormatAttr.TYPE_KEY, ElideGrainTypeFormatAttr.TYPE_MSG)
                .put(ElideJoinTypeFormatAttr.TYPE_KEY, ElideJoinTypeFormatAttr.TYPE_MSG)
                .put(ElideJoinKindFormatAttr.TYPE_KEY, ElideJoinKindFormatAttr.TYPE_MSG)
                .put(ElideTimeFieldTypeFormatAttr.TYPE_KEY, ElideTimeFieldTypeFormatAttr.TYPE_MSG)
                .put(ElideNameFormatAttr.FORMAT_KEY, ElideNameFormatAttr.FORMAT_MSG)
                .put(ElideNameFormatAttr.NAME_KEY, ElideNameFormatAttr.NAME_MSG)
                .put(ElideRSQLFilterFormatAttr.FORMAT_KEY, ElideRSQLFilterFormatAttr.FORMAT_MSG)
                .put(JavaClassNameWithExtFormatAttr.FORMAT_KEY, JavaClassNameWithExtFormatAttr.FORMAT_MSG)
                .put(JavaClassNameFormatAttr.FORMAT_KEY, JavaClassNameFormatAttr.FORMAT_MSG)
```

File: elide-model-config/src/main/java/com/yahoo/elide/modelconfig/jsonformats/ElideNameFormatAttr.java [hunk 4]
```diff
@@ -19,9 +19,11 @@
  * </p>
  */
 public class ElideNameFormatAttr extends AbstractFormatAttribute {
-    private static final String NAME_FORMAT_REGEX = "^[A-Za-z][0-9A-Za-z_]*$";
+    public static final String NAME_FORMAT_REGEX = "^[A-Za-z][0-9A-Za-z_]*$";
```
Comment (wcekan) [inline elide-model-config/src/main/java/com/yahoo/elide/modelconfig/jsonformats/ElideNameFormatAttr.java:22]: Consider pre-formatting the regex.
```suggestion
    public static final Pattern NAME_FORMAT_REGEX = Pattern.compile("^[A-Za-z][0-9A-Za-z_]*$");
```

File: elide-model-config/src/main/java/com/yahoo/elide/modelconfig/jsonformats/ElideNameFormatAttr.java [hunk 5]
```diff
@@ -39,5 +41,9 @@ public void validate(final ProcessingReport report, final MessageBundle bundle,
         if (!input.matches(NAME_FORMAT_REGEX)) {
```
Comment (wcekan) [inline elide-model-config/src/main/java/com/yahoo/elide/modelconfig/jsonformats/ElideNameFormatAttr.java:41]: and use the pattern
```suggestion
        if (!FIELD_NAME_FORMAT_REGEX.matcher(input).matches()) {
```

File: elide-model-config/src/main/java/com/yahoo/elide/modelconfig/jsonformats/ElideArgumentNameFormatAttr.java [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+    public ElideArgumentNameFormatAttr() {
+        super(FORMAT_NAME, NodeType.STRING);
+    }
+
+    @Override
+    public void validate(final ProcessingReport report, final MessageBundle bundle, final FullData data)
+                    throws ProcessingException {
+        final String input = data.getInstance().getNode().textValue();
+
+        if (!input.matches(ElideNameFormatAttr.NAME_FORMAT_REGEX)) {
```
Comment (wcekan) [inline elide-model-config/src/main/java/com/yahoo/elide/modelconfig/jsonformats/ElideArgumentNameFormatAttr.java:39]:
```suggestion
        if (!ElideNameFormatAttr.NAME_FORMAT_REGEX.matcher(input).matches()) {
```

File: elide-model-config/src/main/java/com/yahoo/elide/modelconfig/jsonformats/ValidateArgsPropertiesKeyword.java [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+        }
+
+        @Override
+        public JsonNode digest(final JsonNode schema) {
+            final ObjectNode node = FACTORY.objectNode();
+            if (schema.get(keyword).asBoolean(true)) {
+                node.put(keyword, true);
+            } else {
+                node.put(keyword, false);
+            }
```
Comment (wcekan) [inline elide-model-config/src/main/java/com/yahoo/elide/modelconfig/jsonformats/ValidateArgsPropertiesKeyword.java:72]: Consolidate unnecessary conditional
```suggestion
            node.put(keyword, schema.get(keyword).asBoolean(true));
```

File: elide-model-config/src/main/java/com/yahoo/elide/modelconfig/validator/DynamicConfigValidator.java [hunk 8]
```diff
@@ -720,6 +975,63 @@ private static boolean containsDisallowedWords(String str, String splitter, Set<
                 : Arrays.stream(str.trim().toUpperCase(Locale.ENGLISH).split(splitter)).anyMatch(keywords::contains);
     }
 
+    private void validateArgumentExists(Table currentModel, Named currentColumn, String argName) {
+        // Arguments are not supported for Joins. Maintain a set of required Column arguments, and this set must be
+        // validated for each logical column which references this join.
+        if (currentColumn instanceof Join) {
+            ((Join) currentColumn).getRequiredColumnArgs().add(argName);
```
Comment (aklish) [inline elide-model-config/src/main/java/com/yahoo/elide/modelconfig/validator/DynamicConfigValidator.java:982]: A couple of problems with this:
1. We don't want to put internal validation logic into public contracts (Join configuration class).
2. validation functions should be idempotent as callers will expect them to be stateless.

We need to find another way to do this.  Maybe we need to parse all the join expressions and pull out their required arguments as step one.  At a later step, we can validate that any column which references the join also has all required arguments.  

We also want to make sure we are not duplicating work done by the SQL Reference Table (which does some validation work as well with respect to cycles).  SQL reference table is the only place today that understand the expansion of a join (joins can depend on other joins and so on).  I think we need to rethink how this validation logic responsibility is shared with the SQL Reference table.

File: elide-model-config/src/test/java/com/yahoo/elide/modelconfig/validator/DynamicConfigValidatorTest.java [hunk 9]
```diff
@@ -364,6 +365,66 @@ public void testJoinedTablesDBConnectionNameMismatch() throws Exception {
...
+
+    @Test
+    public void testMissingReferredColumn() throws Exception {
+        String error = tapSystemErr(() -> {
+            int exitStatus = catchSystemExit(() -> DynamicConfigValidator
+                    .main(new String[]{"--configDir", "src/test/resources/validator/invalid_referred_column"}));
+
+            assertEquals(2, exitStatus);
+        });
+        assertEquals("Can't invoke 'revenue' provided in \"sql from='revenueFact' column='revenue[currency:USD]'\" for Column: 'impressionsPerUSD' in Model: 'revenueFact'. Column: 'revenue' is undefined for hjson model: 'revenueFact'\n",
```
Comment (aklish) [inline elide-model-config/src/test/java/com/yahoo/elide/modelconfig/validator/DynamicConfigValidatorTest.java:400]: Models may not come from HJSON.  The returned text should not reference HJSON here.

**Review by aklish — CHANGES_REQUESTED** [review]

**Pull request closed without merging**